clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2591,6 +2544,10 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
 
     const auto *RecordDecl = Ty->getAsCXXRecordDecl();
     createSpecialMethodCall(RecordDecl, getInitMethodName(), BodyStmts);
+    CXXMethodDecl *FinalizeMethod =
+        getMethodByName(RecordDecl, FinalizeMethodName);
+    if (FinalizeMethod)
+      createSpecialMethodCall(RecordDecl, FinalizeMethodName, BodyStmts);
```

Comment:
I think you need to add to `FinalizeStmts`. Not body statements. That way the finalize method calls will all happen at the end when body is constructed.